syntax = "proto3";

package unitedshieldspace;

service UnitedShieldSpace {
    rpc RegisterNewUser(NewUserDetails) returns (UserCreationStatus);
    rpc LoginUser(UserCredentials) returns (LoginResponse);
    rpc GetNewTokens(RefreshTokenDetails) returns (NewTokens);
    rpc UploadFile(stream FileSegment) returns (UploadStatus);
    rpc ListUserFiles(UserDetails) returns (stream FileDetails);
    rpc ListSharedWithMeFiles(UserDetails) returns (stream ExtFileDetails);
    rpc UpdateACL(ACLDetails) returns (ACLUpdateResponse);
}

message NewUserDetails {
    string name = 1;
    string email = 2;
    string password = 3;
}

message UserCreationStatus {
    bool userCreated = 1;
}

message UserCredentials {
    string email = 1;
    string password = 2;
}

message LoginResponse {
    bool loginStatus = 1;
    string name = 2;
    string uid = 3;
    string accessToken = 4;
    string refreshToken = 5;
}

message RefreshTokenDetails {
    string uid = 1;
    string refreshToken = 2;
}

message NewTokens {
    string accessToken = 1;
    string refreshToken = 2;
}

message FileSegment {
    string email = 1;
    string uid = 2;
    string fileName = 3;
    string accessToken = 4;
    bytes fileSegmentData = 5;
}

message UploadStatus {
    bool uploadStatus = 1;
}

message UserDetails {
    string email = 1;
    string accessToken = 2;
}

message FileDetails {
    string name = 1;
    int64 createdOn = 2;
}

message ExtFileDetails {
    string name = 1;
    string owner = 2;
    int64 createdOn = 3;
}

message ACLDetails {
    string owner = 1;
    string name = 2;
    string toEmail = 3;
    bool grant = 4;
    string accessToken = 5;
}

message ACLUpdateResponse {
    bool ACLUpdateStatus = 1;
}